### ISSUE TITLE
Deprecate typesafe variadic class arguments 

### ISSUE THREAD

**Issue by dlangBugzillaToGithub** [dlang/dmd#19890]
### Description

```D
struct A
{
	int a;
}
int f(A a...){ return a.a; }
static assert(f(4)==4); //not callable
```

Typesafe variadic functions are supported with classes, but not with structs. Is there a reason for this limitation?

**Comment by Lars-Kristiansen** [comment]
They are supported, the problem is that you rely on implicit construction of an A instance which does not exist in D.

The correct way to write this is

```d
struct A
{
	int a;
}
int f(A a...){ return a.a; }
static assert(f(A(4))==4); 
```

I think this one can be close as invalid.

**Comment by dkorpel** [comment]
> They are supported, the problem is that you rely on implicit construction of an A instance which does not exist in D.

Implicit construction is supported for class arguments of a typesafe variadic function.

```D
class A
{
    this(int a) {this.a = a;}
	int a;
}
int f(A a...){ return a.a; }
static assert(f(4) == 4);
```

See: https://dlang.org/spec/function.html#typesafe_variadic_functions

> For other types [than arrays and classes], the argument is passed by value.

So yes, the current compiler behavior is correct, but it makes a variadic struct argument completely useless. You might as well just write:

```diff
-int f(A a...){ return a.a; }
+int f(A a){ return a.a; }
```

Asking why arrays and classes are supported here, but not structs, is a valid enhancement request.

However, Walter has tried to deprecate typesafe variadic arguments before, since D has so many different kinds of variadic functions that it's getting really complex. I think most D users don't even know implicit construction of classes exists, I haven't seen them being used. Perhaps instead of adding typesafe variadic structs, we should deprecate typesafe variadic classes instead. (Removing typesafe variadic arrays received some pushback, since it is being used)

**Comment by thewilsonator** [comment]
That is wild! 

> we should deprecate typesafe variadic classes instead.

Yes.

**Comment by Lars-Kristiansen** [comment]
> Implicit construction is supported for class arguments of a typesafe variadic function.

sorry if I mislead people with my attempt to get this issue closed but in first place, why ? usually implicit construction is not a thing.

**Comment by Lars-Kristiansen** [comment]
> Perhaps instead of adding typesafe variadic structs, we should deprecate typesafe variadic classes instead

yes, in a first time, then eventually later type safe variadics.
